k2/csrc/rnnt_decode.h

```diff
@@ -0,0 +1,389 @@
+  // It contains Ragged<ArcInfo> with 2 axes (state, arc).
+  std::vector<std::shared_ptr<Ragged<ArcInfo>>> prev_frames;
+};
+
+class RnntDecodingStreams {
+ public:
+  /* Constructor. Combines multiple RnntDecodingStream objects to create a
+     RnntDecodingStreams object */
+  RnntDecodingStreams(std::vector<std::shared_ptr<RnntDecodingStream>> &srcs,
+                      const RnntDecodingConfig &config);
+
+  /* This function must be called prior to evaluating the joiner network
+     for a particular frame.  It tells the calling code which contexts
+     it must evaluate the joiner network for.
+
+       @param [out] shape  A RaggedShape with 2 axes, representing
+                      [stream][context], will be written to here.
+       @param [out] contexts  An array of shape
+                      [tot_contexts][decoder_history_len], will be output to
+                      here, where tot_contexts == shape->TotSize(1) and
+                      decoder_history_len comes from the config, it represents
+                      the number of symbols in the context of the decode
+                      network (assumed to be finite). It contains the token ids
+                      into the vocabulary(i.e. `0 <= value < vocab_size`).
+  */
+  void GetContexts(RaggedShape *shape, Array2<int32_t> *contexts);
+
+  /*
+    Advance decoding streams by one frame.
+
+      @param [in] logprobs  Array of shape [tot_contexts][num_symbols],
+                    containing log-probs of symbols given the contexts output
+                    by `GetContexts()`. Will satisfy
+                    logprobs.Dim0() == states.TotSize(1).
+   */
+  void Advance(Array2<float> &logprobs);
+
+  /*
+    Generate the lattice.
+
+    Note: The prev_frames_ only contains decoded by current object, in order to
+          generate the lattice we will fisrt gather all the previous frames from
+          individual streams.
+
+      @param [in] num_frames  A vector containing the number of frames we want
+                    to gather for each stream (note: the frames we have
+                    ever received).
+                    It MUST satisfy `num_frames.size() == num_streams_`, and
+                    `num_frames[i] < srcs_[i].prev_frames.size()`.
+      @param [out] ofsa  The output lattice will write to here, its num_axes
+                         equals to 3, will be re-allocated.
+      @param [out] out_map  It it a Array1 with Dim() equals to
+                     ofsa.NumElements() containing the idx01 into the graph of
+                     each individual streams, mapping current arc in ofsa to
+                     original decoding graphs. It may contains -1 which means
+                     this arc is a "termination symbol".
+   */
+  void FormatOutput(std::vector<int32_t> &num_frames, FsaVec *ofsa,
+                    Array1<int32_t> *out_map);
+
+  /*
+    Detach the RnntDecodingStreams, it will update the states & scoers of each
```

Comment:
misspelling: scores.
also appended->append.

Reply:
Detach the RnntDecodingStreams -> Detach the RnntDecodingStreams object.

(This may be my fault, but..) I am concerned that the name Detach() may be too confusable with .detach() in PyTorch.  Perhaps could call it TerminateAndFlushToStreams() ?